learning_observer/learning_observer/learning_tools/canvas.py

```diff
@@ -0,0 +1,162 @@
+import json
+import requests
+import configparser
+
+class Canvas:
+    def __init__(self, config_path='./config.ini'):
+        self.config = configparser.ConfigParser()
+        self.config.read(config_path)
+        self.defaultServer = self.config['CANVAS_CONFIG']['DEFAULT_SERVER']
+        self.access_token = self.config['CANVAS_CONFIG']['ACCESS_TOKEN']
+        self.refresh_token = self.config['CANVAS_CONFIG']['REFRESH_TOKEN']
+        self.client_id = self.config['CANVAS_CONFIG']['CLIENT_ID']
+        self.client_secret = self.config['CANVAS_CONFIG']['CLIENT_SECRET']
+        self.default_version = 'v1'
+        self.defaultPerPage = 10000
+        self.base_url = f'https://{self.defaultServer}/api/{self.default_version}'
+
+    def update_access_tokens(self, access_token):
+        self.config['CANVAS_CONFIG']['ACCESS_TOKEN'] = access_token
+        self.access_token = access_token
+        with open('./config.ini', 'w') as configfile:
+            self.config.write(configfile)
+
+    def api_call(self, method, endpoint, params=None, data=None, absolute_url=False, retry=True):        
+        if absolute_url:
+            url = endpoint
+        else:
+            url = self.base_url + endpoint
+            # Append params key/values if added
+            if params:
+                url += '?' + '&'.join(f"{k}={v}" for k, v in params.items())
+
+        headers = {
+            'Authorization': f'Bearer {self.access_token}',
+            'Content-Type': 'application/json'
+        }
+
+        response_func = getattr(requests, method.lower())
+        response = response_func(url, headers=headers, params=params, data=json.dumps(data))
+        
+        # Handle invalid/expired token
+        if response.status_code == 401 and retry:
+            new_tokens = self.refresh_tokens()
+            if 'access_token' in new_tokens:
+                self.update_access_tokens(new_tokens['access_token'])
+                return self.api_call(method, endpoint, params, data, absolute_url, retry=False)
+        
+        if response.status_code != 200:
+            response.raise_for_status()
+        
+        return response.json()
+
+    def refresh_tokens(self):
+        url = f'https://{self.defaultServer}/login/oauth2/token'
+        params = {
+            "grant_type": "refresh_token",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "refresh_token": self.refresh_token
+        }
+        return self.api_call('POST', url, params=params, absolute_url=True)
+
+    def get_account(self):
+        return self.api_call('GET', '/accounts/self')
```

Comment:
The google file utilized a dictionary of urls to help cut down on repeated code - this method and the following ones all look super similar. Perhaps that is feasible here as well.